core/trino-main/src/main/java/io/trino/operator/window/matcher/Matcher.java

```diff
@@ -13,89 +13,100 @@
  */
 package io.trino.operator.window.matcher;
 
+import com.google.common.collect.ArrayListMultimap;
+import com.google.common.collect.ListMultimap;
+import io.airlift.slice.SizeOf;
+import io.trino.memory.context.LocalMemoryContext;
 import io.trino.operator.window.pattern.LabelEvaluator;
+import org.openjdk.jol.info.ClassLayout;
+
+import java.util.List;
 
-import static com.google.common.base.Preconditions.checkState;
 import static io.trino.operator.window.matcher.MatchResult.NO_MATCH;
 
 public class Matcher
 {
     private final Program program;
+    private final ThreadEquivalence threadEquivalence;
 
     private static class Runtime
     {
-        private int step;
+        private static final int INSTANCE_SIZE = ClassLayout.parseClass(Runtime.class).instanceSize();
+        private static final int INTEGER_SIZE = ClassLayout.parseClass(Integer.class).instanceSize();
+
+        // a helper structure for identifying equivalent threads
+        // program pointer (instruction) --> list of threads that have reached this instruction
+        private ListMultimap<Integer, Integer> threadsAtInstructions = ArrayListMultimap.create();
```

Comment:
Why not `IntListMap`?

Reply:
I purposefully used `IntListMap` only in the cases where `threadId`s are the keys. This is because the `threadId`s are reused, and so they are supposed to occupy a more or less consistent prefix of the table.
In the structure `threadsAtInstructions`, instruction indexes are the keys. I didn't want to have a table where many positions are not used.
Do you think it is more efficient to use `IntListMap` here?

Reply:
Given that the keys (instructions) are going to be fairly small cardinality in general, it doesn't matter much. The amount of space wasted by a few unused keys is likely going to be much smaller than the overhead from a ListMultimap and boxed integers.